http/src/main/scala/org/http4s/blaze/http/http2/WriteControllerImpl.scala

```diff
@@ -0,0 +1,154 @@
+package org.http4s.blaze.http.http2
+
+import java.nio.ByteBuffer
+import java.util
+
+import org.http4s.blaze.pipeline.TailStage
+import org.log4s.getLogger
+
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{Future, Promise}
+import scala.util.{Failure, Success}
+import scala.util.control.NonFatal
+
+/** Gracefully coordinate writes
+  *
+  * The `WriteController`s job is to direct outbound data in a fair, efficient, and
+  * thread safe manner. All calls to the `WriteController` are expected to come from
+  * within the session executor.
+  *
+  * @param highWaterMark number of bytes that will trigger a flush.
+  */
+private final class WriteControllerImpl(
+  session: SessionCore,
+  highWaterMark: Int,
+  tailStage: TailStage[ByteBuffer]
+) extends WriteController {
+  import WriteControllerImpl._
+
+  private[this] val logger = getLogger
+  // TODO: if we implement priorities, we should prioritize writes as well.
+  private[this] val interestedStreams = new java.util.ArrayDeque[WriteInterest]
+  private[this] val pendingWrites = new util.ArrayDeque[Seq[ByteBuffer]]
+
+  private[this] var state: State = Idle
+
+  def close(): Future[Unit] = state match {
+    case Idle =>
+      state = Closed
+      Future.successful(())
+
+    case Flushing =>
+      val p = Promise[Unit]
+      state = Closing(p)
+      p.future
+
+    case Closing(p) =>
+      p.future
+
+    case Closed =>
+      Future.successful(())
+  }
+
+  private[this] def pendingInterests: Boolean = !pendingWrites.isEmpty || !interestedStreams.isEmpty
+
+  def write(data: Seq[ByteBuffer]): Boolean = state match {
+    case Idle | Flushing | Closing(_) =>
+      pendingWrites.addLast(data)
+      maybeWrite()
+      true
+
+    case Closed =>
+      false
+  }
+
+  def write(data: ByteBuffer): Boolean = write(data::Nil)
```

Comment:
Since you like to save every nanosecond, is that the fastest `Seq` you can create?  I think I've seen this a lot of places.  (I don't know, but it seems like the sort of thing you might.)

Reply:
I think it's about as good as you can get for a single element: should just be a single `::`. I'm interested if you know of a better one.